appserver/payara-appserver-modules/hazelcast-ejb-timer/src/main/java/fish/payara/ejb/timer/hazelcast/HazelcastTimerStore.java

```diff
@@ -1193,4 +1157,37 @@ private boolean restoreEJBTimers() {
         return rc;
     }
 
+    @Override
+    public void memberAdded(MemberEvent event) {
+        //do nothing
+    }
+
+    @Override
+    public void memberRemoved(MemberEvent event) {
+        ILock hazelcastLock = Globals.getDefaultBaseServiceLocator().getService(HazelcastCore.class).getInstance().getLock("EJB-TIMER-LOCK");
+        hazelcastLock.lock();
+        try {
+            Collection<HZTimer> allTimers = pkCache.values();
+            Collection<HZTimer> removedTimers = new HashSet<>();
+            for (HZTimer timer : allTimers) {
+                if (timer.getMemberName().equals(event.getServer())) {
+                    removedTimers.add(timer);
+                }
+            }
+            
+            
+            
+            if (!removedTimers.isEmpty()) {
+                logger.log(Level.INFO, "==> Restoring Timers ... ");
+                Collection<HZTimer> restored = _restoreTimers(removedTimers);
+                for (HZTimer timer : restored) {
+                    pkCache.putAsync(timer.getKey().getTimerId(), timer);
+                }
+                logger.log(Level.INFO, "<== ... Timers Restored.");
```

Comment:
This could become a bit confusing in the logs as it suggest that at this point the timers are restored which might not yet be the case due to `putAsync`.